Feature/bbs 38 ding serendipity part i

danquah opened the pull request.

kasperg commented on modules/ding_availability/ding_availability.module:
```diff
@@ -164,7 +164,7 @@ function ding_availability_holdings($provider_ids) {
 
     // Marks internet resources as available.
-    if ($item['is_internet']) {
+    if (!empty($item['is_internet'])) {
```
:+1:

kasperg commented on modules/ding_provider/connie_search/connie_search.test:
```diff
@@ -128,4 +128,18 @@ class ConnieSearchSearchProviderImplementationTestCase extends DingSearchProvide
+   *   List of fields.
+   */
+  public function getExpectedCommonFields() {
+    $class = new ReflectionClass('\Ting\Search\TingSearchCommonFields');
```
I would prefer `TingSearchCommonFields::class` and then `use` the class with the full namespace at the top of the file. Less stringwork.

danquah replied:
👍

kasperg commented on modules/ding_provider/connie_search/connie_search.test:
```diff
@@ -128,4 +128,18 @@ class ConnieSearchSearchProviderImplementationTestCase extends DingSearchProvide
+   */
+  public function getExpectedCommonFields() {
+    $class = new ReflectionClass('\Ting\Search\TingSearchCommonFields');
+    // Return all static properties that does not start with _.
```
Why do we not want to return static properties starting with `_`?

kasperg replied:
Another approach would be to let the "Enum" provide a function which will return all legal values. I actually prefer that.

danquah replied:
I have a single _ field in there for holding the prefix for the fields - but I like the idea of just having the class itself enumerate the list of possible values way better, it will make a number of other cases much cleaner.

kasperg commented on modules/p2/ding_serendipity/ding_serendipity.module:
```diff
@@ -451,8 +455,10 @@ function ding_serendipity_user_audience($account = NULL) {
   }
 
+  $query = ting_start_query();
+  $query->setMaterialFilter($ids);
```
Using a fluid API here would be nice: 

```php
$query = ting_start_query()->setMaterialFilter($ids);

danquah replied:
👍

kasperg commented on modules/p2/ding_serendipity/ding_serendipity.module:
```diff
@@ -973,6 +979,60 @@ function ding_serendipity_do_search($query, array $options = array()) {
+
+  // Create result array, stop if count hits limit.
+  $count = 0;
+  foreach ($collections as $collection) {
```
Perhaps change to below to leave out ugly `if/break`.

```php
foreach (array_slice(0, $limit, $collections) as $collection) {

kasperg commented on modules/p2/ding_serendipity_fallback/ding_serendipity_fallback.module:
```diff
@@ -44,7 +44,12 @@ function ding_serendipity_fallback_ting_object_serendipity_add($context) {
     try {
-      $results = ding_serendipity_do_search($query, array('limit' => 8));
+      /** @var \Ting\Search\TingSearchRequest $sal_query */
+      $sal_query = ting_start_query();
```
Is the DocBlock needed with proper type hinting for `ting_start_query()`?

danquah replied:
Nope, it was a leftover, removing it.

kasperg commented on modules/p2/ding_serendipity_lists/ding_serendipity_lists.module:
```diff
@@ -356,8 +361,10 @@ function ding_serendipity_lists_author_from_lists_serendipity_add($context) {
+    $field_filters = [];
     // Pick the first author.
     $terms[] = 'facet.creator="' . $author[0] . '"';
+    $field_filters[] = new TingSearchFieldFilter(TingSearchCommonFields::AUTHOR, $author[0], TingSearchFieldFilter::OP_AND);
```
Do we need `$terms` above now that we have `$field_filters`?

danquah replied:
Nope, leftover yet again.

kasperg commented on modules/p2/ding_serendipity_lists/ding_serendipity_lists.module:
```diff
@@ -370,24 +377,26 @@ function ding_serendipity_lists_author_from_lists_serendipity_add($context) {
       // Fallback to dc.subject.
       if ($value === FALSE) {
         $value = 'facet.subject="' . $term->name . '"';
+        $field_filters[] = new TingSearchFieldFilter(TingSearchCommonFields::SUBJECT, $term->name, TingSearchFieldFilter::OP_AND);
```
Do we need `$value` above now that we have `$field_filters`?

danquah replied:
nope

kasperg commented on modules/ting/src/Search/BooleanStatementInterface.php:
```diff
@@ -0,0 +1,39 @@
+namespace Ting\Search;
+
+/**
+ * Interface BooleanStatementInterface
```
IMO this line is unecessary.

danquah replied:
👍

kasperg commented on modules/ting/src/Search/TingSearchRequest.php:
```diff
@@ -45,12 +45,75 @@ class TingSearchRequest {
+
+  protected $fieldFilters;
+
+  protected $populateCollections = FALSE;
```
These might need some documentation.

danquah replied:
👍

kasperg commented on modules/ting_material_details/ting_material_details.field.inc:
```diff
@@ -478,7 +478,7 @@ function ting_material_details_field_formatter_settings_summary($field, $instanc
  */
 function ting_material_details_get_object_property($field_name, TingEntity $entity) {
-  if ($entity->reply === NULL) {
+  if ($entity->reply === NULL || $entity->reply === FALSE) {
```
Perhaps even `if (empty($entity->reply) { {`?

danquah replied:
👍

kasperg commented on modules/ting/src/Search/TingSearchRequest.php:
```diff
@@ -124,13 +207,121 @@ public function getFacets() {
+   * Filter(s) or statement group(s) to add to the query.
+   *
+   * @param BooleanStatementInterface[]|BooleanStatementInterface $filters
+   *   One or more filters.
```
:+1: Nice DX

kasperg commented on modules/ting/src/Search/TingSearchRequest.php:
```diff
@@ -124,13 +207,121 @@ public function getFacets() {
+  }
+
+  /**
+   * Set the list of material ids the query is constrained to.
```
Do we have any reference to where one might find such an id? A reference to a method on a TingObject perhaps?

danquah replied:
👍 adding comments.

kasperg commented on modules/ting/src/Search/TingSearchRequest.php:
```diff
@@ -45,12 +45,75 @@ class TingSearchRequest {
+   *
+   * @var string
+   */
+  protected $rawQuery;
```
I am in general a bit uncertain what the "raw" part of the query actually means. Could that be explained better?

danquah replied:
I have documented the setter a bit more instead.

kasperg commented on modules/ting/src/Search/TingSearchFieldFilter.php:
```diff
@@ -0,0 +1,110 @@
+   *   Operator to use when comparing the evaluated field with a previous
+   *   statement.
+   */
+  public function __construct($name, $value, $operator = '=', $logic_operator = BooleanStatementInterface::OP_AND) {
```
Should operators like `=` be enumerated as well?

danquah replied:
Might be a good idea, can't really decide whether it will be going a bit far or make the abstraction layer more robust - I'll punt it for now but will leave a comment.

xendk replied:
Please no. Using strings mimics the Drupal query classes (SQL/EntityFieldQuery), which is good.

kasperg replied:
The current enum implementation is in essence stringly typed anyway. If we continue on that route I see no harm in at least having an operator enum and validating the input. This lets developers choose whether they want to use `Operator::EQUALS` or `=`.

kasperg commented on modules/ting/src/Search/StatementGroupRender.php:
```diff
@@ -0,0 +1,162 @@
+   * @param \Ting\Search\BooleanStatementGroup $group
+   *   The group to be rendered.
+   *
+   * @param string                             $rendered_statement
```
Odd spacing here :smile:

danquah replied:
_really_ need to fix that phpstom setting :)

kasperg commented on modules/ting/src/Search/BooleanStatementGroup.php:
```diff
@@ -0,0 +1,69 @@
+   * @param string                               $logic_operator
+   *   A TingSearchBooleanStatementInterface::OP_* operation.
+   */
+  public function __construct(array $statements, $logic_operator = 'AND') {
```
Change param to `$logic_operator = BooleanStatementInterface::OP_AND`?

danquah commented:
@kasperg I've addressed most of the comments, and have converted the statementgroup class to be instancebased instead of static and have added a unittest
